tests/components/zimi/test_config_flow.py

```diff
@@ -0,0 +1,219 @@
+    assert result["title"] == "ZIMI Controller"
+    assert result["data"] == {
+        "title": "ZIMI Controller",
+        "host": test_host,
+        "port": test_port,
+        "timeout": 3,
+        "verbosity": 1,
+        "watchdog": 1800,
+        "mac": format_mac(test_mac),
+    }
+
+
+async def test_successful_config_discovery(
+    hass: HomeAssistant, discovery_service
+) -> None:
+    """Test successful configuration with automatic discovery."""
+    test_mac = "AA:BB:CC:DD:EE:FF"
+    discovered_host = "192.168.1.200"
+    discovered_port = 5003
+
+    discovery_description = MagicMock()
+    discovery_description.host = discovered_host
+    discovery_description.port = discovered_port
+    discovery_service.discover.return_value = discovery_description
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {
+            CONF_HOST: "",  # Empty host triggers discovery
+            CONF_PORT: 5003,
+            MAC: test_mac,
+        },
+    )
+
+    assert result["type"] == data_entry_flow.FlowResultType.CREATE_ENTRY
+    assert result["data"]["host"] == discovered_host
+    assert result["data"]["port"] == discovered_port
+
+
+async def test_discovery_failure(hass: HomeAssistant, discovery_service) -> None:
+    """Test failed discovery."""
+    discovery_service.discover.side_effect = ControlPointError("Discovery failed")
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {
+            CONF_HOST: "",
+            CONF_PORT: 5003,
+            MAC: "AA:BB:CC:DD:EE:FF",
+        },
+    )
+
+    assert result["type"] == data_entry_flow.FlowResultType.FORM
+    assert result["errors"] == {"base": "discovery_failure"}
+
+
+async def test_invalid_host(hass: HomeAssistant, socket_mock) -> None:
+    """Test error on invalid host."""
+    socket_mock.gethostbyname.side_effect = socket.herror
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {
+            CONF_HOST: "invalid_host",
+            CONF_PORT: 5003,
+            MAC: "AA:BB:CC:DD:EE:FF",
+        },
+    )
+
+    assert result["type"] == data_entry_flow.FlowResultType.FORM
+    assert result["errors"] == {"base": "invalid_host"}
+
+
+async def test_connection_refused(hass: HomeAssistant, socket_mock) -> None:
+    """Test error on connection refused."""
+    socket_mock.gethostbyname.return_value = "192.168.1.100"
+    socket_mock.socket.return_value.connect.side_effect = ConnectionRefusedError
```

Comment:
the three look like they can be parametrized, also make sure they all come to a terminal state (`CREATE_ENTRY` or `ABORT`) by resolving the errors, to show the flow can recover

Reply:
Done